### PR TITLE
Desktop nrf54h

### PR DESCRIPTION
**Pull request by zycz**
Add configuration for nrf54h20 for nrf_desktop application

### CODE REVIEW

**Comment by kapi-no** [comment]
I had some time to look into the Bluetooth initialization issue with the nRF Desktop app that @zycz mentioned before going on vacation leave.

```
[00019545] <inf> app_event_manager: e:module_state_event module:main state:READY
[00019681] <inf> hids: Service initialized
[00019730] <inf> hid_state: Init HID state!
[01019953] <err> bt_hci_driver: Endpoint binding failed with -11
[01019987] <err> bt_hci_core: HCI driver open failed (-11)
[01019998] <err> ble_state: Cannot initialize
```
The error here stems from the fact that we end up in a deadlock and the waiting semaphore times out (`bt_hci_driver: Endpoint binding failed with -11`). We register the IPC endpoint and wait for endpoint binding to finalize the Bluetooth HCI driver open operation. The waiting occurs in the System Workqueue context. However, the ICBMsg backend from the IPC service schedules a system work during the endpoint registration, in which it finalizes the binding operation (also in the System Workqueue context). As the Bluetooth HCI driver with its wait operation keeps the System Workqueue context busy - the endpoint binding cannot be completed by the ICBMsg backend before the HCI driver semaphore timeout.

See the following branch and commit that works around the issue in a dirty way:

https://github.com/kapi-no/sdk-zephyr/tree/nrf_desktop_nrf54h20_ipc_test
https://github.com/kapi-no/sdk-zephyr/commit/e9e86848f304c5ba717d58ad2cb4d3812d3e3ecb

We need to come up with a proper solution on how to resolve this issue.

**Comment by kapi-no** [comment]
> I had some time to look into the Bluetooth initialization issue with the nRF Desktop app that @zycz mentioned before going on vacation leave.
> 
> ```
> [00019545] <inf> app_event_manager: e:module_state_event module:main state:READY
> [00019681] <inf> hids: Service initialized
> [00019730] <inf> hid_state: Init HID state!
> [01019953] <err> bt_hci_driver: Endpoint binding failed with -11
> [01019987] <err> bt_hci_core: HCI driver open failed (-11)
> [01019998] <err> ble_state: Cannot initialize
> ```
> 
> The error here stems from the fact that we end up in a deadlock and the waiting semaphore times out (`bt_hci_driver: Endpoint binding failed with -11`). We register the IPC endpoint and wait for endpoint binding to finalize the Bluetooth HCI driver open operation. The waiting occurs in the System Workqueue context. However, the ICBMsg backend from the IPC service schedules a system work during the endpoint registration, in which it finalizes the binding operation (also in the System Workqueue context). As the Bluetooth HCI driver with its wait operation keeps the System Workqueue context busy - the endpoint binding cannot be completed by the ICBMsg backend before the HCI driver semaphore timeout.
> 
> See the following branch and commit that works around the issue in a dirty way:
> 
> https://github.com/kapi-no/sdk-zephyr/tree/nrf_desktop_nrf54h20_ipc_test [kapi-no/sdk-zephyr@e9e8684](https://github.com/kapi-no/sdk-zephyr/commit/e9e86848f304c5ba717d58ad2cb4d3812d3e3ecb)
> 
> We need to come up with a proper solution on how to resolve this issue.

Today, we discussed strategy for handling this issue and prepared an interim solution to work around the issue:

https://github.com/nrfconnect/sdk-zephyr/pull/1676

**Comment by kapi-no** [comment]
The IPC issue is now fixed in the main sdk-nrf branch and we can ignore my previous comments

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -212,6 +212,9 @@ nRF Desktop
     The PDK can act as a sample mouse or keyboard.
     It supports the Bluetooth LE HID data transport and uses SoftDevice Link Layer with Low Latency Packet Mode (LLPM) enabled.
     The PDK uses MCUboot bootloader built in the direct-xip mode (``MCUBOOT+XIP``) and supports firmware updates using the :ref:`nrf_desktop_dfu`.
+  * Support for the nRF54H20 DK with the ``nrf54h20dk/nrf54h20/cpuapp`` board target.
```
Comment (kapi-no) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:215]: It would be best either to update the documentation later in one go or add it gradually with each feature (this PR introduces BLE and LLPM).

If you choose the second option (adding it gradually), there are still some missing stuff here (e.g. the nRF Desktop Readme file update).

File: applications/nrf_desktop/sysbuild/hci_ipc/prj.conf
Comment (kapi-no) [inline applications/nrf_desktop/sysbuild/hci_ipc/prj.conf:1]: How do we distinguish between the HCI IPC configurations for each target?

Comment (kapi-no) [reply applications/nrf_desktop/sysbuild/hci_ipc/prj.conf:1]: Can you align with the @nordicjm approach in his sysbuild treewide PR?

https://github.com/nrfconnect/sdk-nrf/pull/13153/files#diff-1643e8fdce13f0c8d003f47f5bf231ddfe83c08c8f5b3436c75b7d70f96f28b1R1

Comment (kapi-no) [reply applications/nrf_desktop/sysbuild/hci_ipc/prj.conf:1]: CC: @mkapala-nordic

**Review by annwoj — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -212,6 +212,9 @@ nRF Desktop
...
+    The PDK can act as a sample mouse.
+    It supports the Bluetooth LE HID data transport and uses SoftDevice Link Layer with Low Latency Packet Mode (LLPM) enabled.
```
Comment (annwoj) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:217]:
```suggestion
  * Support for the nRF54H20 DK with the ``nrf54h20dk/nrf54h20/cpuapp`` board target.
    The DK can act as a sample mouse.
    It supports the Bluetooth LE HID data transport and uses SoftDevice Link Layer with Low Latency Packet Mode (LLPM) enabled.
```
overalll lgtm, but please add it to the cs2 release notes: https://github.com/nrfconnect/sdk-nrf/pull/15273

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 3]
```diff
@@ -6,6 +6,23 @@
 target_sources_ifdef(CONFIG_DESKTOP_HWID
 		     app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/hwid.c)
 
+if(CONFIG_HWINFO_NRF AND CONFIG_BOARD_NRF54H20DK_NRF54H20_CPUAPP)
+         message(FATAL_ERROR
+           "CONFIG_HWINFO_NRF is currently not supported on nrf54h20dk_nrf54h20_cpuapp.")
+       endif()
```
Comment (kapi-no) [inline applications/nrf_desktop/src/util/CMakeLists.txt:12]: alignment

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 4]
```diff
@@ -6,6 +6,23 @@
...
+       
+       if(CONFIG_DESKTOP_HWINFO_BLE_ADDRESS_FICR)
```
Comment (kapi-no) [inline applications/nrf_desktop/src/util/CMakeLists.txt:14]: alignment

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 5]
```diff
@@ -6,6 +6,23 @@
 target_sources_ifdef(CONFIG_DESKTOP_HWID
 		     app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/hwid.c)
 
+if(CONFIG_HWINFO_NRF AND CONFIG_BOARD_NRF54H20DK_NRF54H20_CPUAPP)
```
Comment (kapi-no) [inline applications/nrf_desktop/src/util/CMakeLists.txt:9]: Could you migrate the HWinfo part as a dedicated commit (as it was originally contributed)? We may want to revert it at some point.

BTW, is this change still needed? If the FICR registers are populated we can remove this workaround.

File: applications/nrf_desktop/sample.yaml [hunk 6]
```diff
@@ -171,3 +171,18 @@ tests:
...
+    integration_platforms:
+      - nrf54h20dk/nrf54h20/cpuapp
+  applications.nrf_desktop.zrelease_nrf54h20:
+    build_only: true
+    sysbuild: true
+    platform_allow: >
+      nrf54h20dk/nrf54h20/cpuapp
+    integration_platforms:
+      - nrf54h20dk/nrf54h20/cpuapp
+    extra_args:
```
Comment (kapi-no) [inline applications/nrf_desktop/sample.yaml:188]: You are missing suffix configuration here

**Review by kapi-no — COMMENTED** [review]
This is a high-level review. I focused mainly on the overall file structure and the required components for the nRF Desktop support on the nRF54H20 DK target.

Please address compliance issues / GitHub issues with formatting

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 2 above)
Comment (zycz) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:217]: How should I add it to file which is not merged yet?

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 1 above)
Comment (zycz) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:215]: I thought about adding all documentation at once later when more features are merged.

File: applications/nrf_desktop/sysbuild/hci_ipc/prj.conf
Comment (MarekPieta) [reply applications/nrf_desktop/sysbuild/hci_ipc/prj.conf:1]: Since we use `hci_ipc` also for nRF53, we should be able to provide separate configurations per board

Comment (MarekPieta) [reply applications/nrf_desktop/sysbuild/hci_ipc/prj.conf:1]: @nordicjm seems to use `sysbuild.cmake` and `configuration/board_name/images` directory to define configuration per board. The solution may also fix issues of this PR.

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 7]
```diff
@@ -6,6 +6,23 @@
...
+         message(FATAL_ERROR
+           "CONFIG_HWINFO_NRF is currently not supported on nrf54h20dk_nrf54h20_cpuapp.")
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/util/CMakeLists.txt:11]: `nrf54h20dk/nrf54h20/cpuapp`

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+CONFIG_RESET_ON_FATAL_ERROR=n
+
+CONFIG_SPEED_OPTIMIZATIONS=y
+
+CONFIG_REBOOT=y
+
+CONFIG_GPIO=y
+
+CONFIG_LED=y
+# PWM not supported on this target, using GPIO instead.
```
Comment (MarekPieta) [inline applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:83]: PWM seems already supported for nRF54H20. I would suggest to open a follow-up ticket to enable it.

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: applications/nrf_desktop/src/util/Kconfig [hunk 9]
```diff
@@ -13,6 +13,26 @@ config DESKTOP_HWID
...
+	default 0x0b1e
+	depends on DESKTOP_HWINFO_BLE_ADDRESS_FICR
+	help
+	  Postfix appended in big-endian to the HW ID derived from the BLE address.
+
+config HWINFO_NRF
+	bool
+	default n if BOARD_NRF54H20DK_NRF54H20_CPUAPP
+	help
+	  The nRF HWINFO driver does not support nRF54H20 yet.
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/util/Kconfig:34]: There seem to be no need to explicitly disable it now. It is no longer used for nRF54H.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 2 above)
Comment (annwoj) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:217]: Please add the entries applicable for the 2.6.99-cs2 release via a comment in Francesco's PR. This is a general remark. 
Release notes are always the last ones to get merged, so naturally the content needs to be updated prior to that.
Due to the specific nature of this release, we have a new changelog file with nRF54H20-specific entries, similarly to what we've done for nRF54L15.

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+CONFIG_CAF_BUTTONS=y
+CONFIG_CAF_BUTTONS_POLARITY_INVERSED=y
+CONFIG_CAF_BUTTONS_PM_KEEP_ALIVE=n
+
+CONFIG_CAF_CLICK_DETECTOR=y
+
+CONFIG_CAF_LEDS=y
+
+CONFIG_CAF_FORCE_POWER_DOWN_EVENTS=y
```
[the remark this answers is not in the document]
Comment (zycz) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:33]: It fails build without it

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP=n
+CONFIG_LOG_MODE_DEFERRED=y
+CONFIG_LOG_PRINTK=y
+CONFIG_LOG_BUFFER_SIZE=4096
+CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=1024
+
+################################################################################
+# DFU configuration
+
+CONFIG_DESKTOP_CONFIG_CHANNEL_DFU_ENABLE=n
```
[the remark this answers is not in the document]
Comment (zycz) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:116]: We don't have such option or am I missing something.

[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:116]: I meant you can remove these 3 lines from the file :)

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:33]: I guess, we refer to the event without `IS_ENABLED(CONFIG_CAF_FORCE_POWER_DOWN_EVENTS)`. Then I would suggest to add the missing `IS_ENABLED(CONFIG_CAF_FORCE_POWER_DOWN_EVENTS)`, instead of working it around in the configuration (should be really simple to fix it properly in code).

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf (hunk 8 above)
Comment (MarekPieta) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:83]: Have you opened the ticket already? Please link opened ticket in similar case to simplify finding it

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf (hunk 11 above)
[the remark this answers is not in the document]
Comment (zycz) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:116]: Ok, it was turned on by config in the same file :)

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 1 above)
Comment (kapi-no) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:215]: so please remove the changelog item as this approach is inconsistent

Comment (zycz) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:215]: So we shouldn't add entry in changelogs if we do not add doc support?

File: applications/nrf_desktop/src/util/Kconfig (hunk 9 above)
Comment (zycz) [reply applications/nrf_desktop/src/util/Kconfig:34]: Are you sure?

If I remove this commit I get:

"ASSERTION FAIL [ret == 8] @ WEST_TOPDIR/nrf/applications/nrf_desktop/src/util/hwid.c:16
[54218097] <err> os: r0/a1:  0x00000004  r1/a2:  0x00000010  r2/a3:  0x00000005
[54218101] <err> os: r3/a4:  0x2f011ca8 r12/ip:  0x3fffffff r14/lr:  0x0e0aa55f
[54218104] <err> os:  xpsr:  0x41000000
[54218106] <err> os: Faulting instruction address (r15/pc): 0x0e0ac6b0
[54218127] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[54218136] <err> os: Current thread: 0x2f012e98 (unknown)
[54307524] <err> os: Halting system"

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 1 above)
Comment (kapi-no) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst]: yes, as it is part of doc support. Either you cover everything or add it later as part of a follow-up ticket.

File: applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf (hunk 8 above)
Comment (zycz) [reply applications/nrf_desktop/configuration/nrf54h20dk_nrf54h20_cpuapp/prj.conf:80]: https://nordicsemi.atlassian.net/browse/NCSDK-27489

File: applications/nrf_desktop/sample.yaml [hunk 12]
```diff
@@ -171,3 +171,18 @@ tests:
...
+    integration_platforms:
+      - nrf54h20dk/nrf54h20/cpuapp
+  applications.nrf_desktop.zrelease_nrf54h20:
+    build_only: true
+    sysbuild: true
+    platform_allow: >
+      nrf54h20dk/nrf54h20/cpuapp
+    integration_platforms:
+      - nrf54h20dk/nrf54h20/cpuapp
+    extra_args: FILE_SUFFIX="release"
```
Comment (kapi-no) [inline applications/nrf_desktop/sample.yaml:188]: You don't really need to use quotation marks

```suggestion
    extra_args: FILE_SUFFIX=release
```

File: applications/nrf_desktop/src/util/Kconfig (hunk 9 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/util/Kconfig:34]: Is any hwinfo driver actually compiled in after you remove the commit ?

**Review by mkapala-nordic — COMMENTED** [review]
Beside compliance failures related to commit messages and already present suggestions it looks fine to me.

File: applications/nrf_desktop/src/util/Kconfig (hunk 9 above)
Comment (zycz) [reply applications/nrf_desktop/src/util/Kconfig:34]: The test passes because it uses the dummy implementation unit for the HWinfo implementation:
hwinfo_weak_impl.c

This backend implements the HW INfo driver API by returning -ENOSYS error:

File: subsys/caf/modules/ble_adv.c [hunk 13]
```diff
@@ -888,7 +888,9 @@ static bool handle_ble_peer_event(const struct ble_peer_event *event)
 				    (state != STATE_OFF)) {
 					LOG_INF("Bonded peer power off, force system power down");
 					update_state(STATE_OFF);
-					force_power_down();
+					if (IS_ENABLED(CONFIG_CAF_FORCE_POWER_DOWN_EVENTS)) {
```
Comment (MarekPieta) [inline subsys/caf/modules/ble_adv.c:891]: This is not a proper fix. We need to add missing dependencies to `CONFIG_CAF_BLE_ADV_POWER_DOWN_ON_DISCONNECTION_REASON_0X15` in nRF Desktop.

`depends on DESKTOP_BLE_ADV(=y)` -> `depends on CAF_BLE_ADV_PM_EVENTS(=n) && CAF_BLE_ADV(=y) && CAF(=y)`
Otherwise we can have `CONFIG_CAF_BLE_ADV_POWER_DOWN_ON_DISCONNECTION_REASON_0X15` enabled without `CONFIG_CAF_FORCE_POWER_DOWN_EVENTS` (should not happen)

Comment (MarekPieta) [reply subsys/caf/modules/ble_adv.c:891]: Seems that we missed the fact that `default y` in overlay can be applied even if dependencies are not met.

File: applications/nrf_desktop/src/util/Kconfig (hunk 9 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/util/Kconfig:34]: btw. Now I think there was a bit of misunderstanding here. I just meant to remove the `default n` for `HWINFO_NRF` that is introduced here as it's no longer needed.

**Review by MarekPieta — CHANGES_REQUESTED** [review]

**Comment by kapi-no** [comment]
Requires the following change to pass the Twister stage:

https://github.com/nrfconnect/sdk-nrf/pull/15346

or the upmerge PR (that contains this fix:

https://github.com/nrfconnect/sdk-nrf/pull/15351

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 14]
```diff
@@ -6,6 +6,17 @@
 target_sources_ifdef(CONFIG_DESKTOP_HWID
 		     app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/hwid.c)
 
+if(CONFIG_DESKTOP_HWINFO_BLE_ADDRESS_FICR)
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/util/CMakeLists.txt:9]: You can slightly update commit message here (you no longer need to disable the `nRF hwinfo` driver).

**Review by MarekPieta — APPROVED** [review]

**Review by kapi-no — APPROVED** [review]

File: applications/nrf_desktop/src/util/CMakeLists.txt [hunk 15]
```diff
@@ -6,6 +6,17 @@
...
+         message(WARNING
+           "CONFIG_DESKTOP_HWINFO_BLE_ADDRESS_FICR is enabled. "
+           "Devices with this HWINFO driver are using BLE address from the FICR register "
+           "expanded with a value defined in the CONFIG_DESKTOP_HWINFO_BLE_ADDRESS_FICR_POSTFIX "
+           "Kconfig option. If this register is not populated (as on nRF54H20-PDK ES3.5), "
```
Comment (mkapala-nordic) [inline applications/nrf_desktop/src/util/CMakeLists.txt:14]: `(as on nRF54H20-PDK ES3.5)` should be removed

**Review by mkapala-nordic — APPROVED** [review]
